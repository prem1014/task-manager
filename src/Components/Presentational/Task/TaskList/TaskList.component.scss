.task-list {
    margin: 10px 0;
    padding: 20px;
    box-shadow: 0 3px 10px grey;
    background-color: #f3f3f3;

    .task-card{
        border: 1px solid #c2c2c2;
        padding: 10px;
        background-color: white;
    }

    h5 {
        display: inline-block;
    }
}