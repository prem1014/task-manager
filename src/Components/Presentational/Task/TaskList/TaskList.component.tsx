import React, { useState } from 'react';

import './TaskList.component.scss';
import Button from '../../../Shared/Form-Elements/Button/Button.component';
import AppModal from '../../../Shared/Modal/Modal.component';
import TaskCardForm from '../Task-Card-Form/Task-Card-Form.component';
import RemoveIcon from '../../../Shared/Remove-Icon/Remove-Icon.component';

interface ITaskListProps {
    taskListId: number;
    name: string;
    taskCard: any;
    taskCardCreated(taskListId: number, cardName: string): any
    onTaskCardDropped(draggedFromtaskListId: number, cardId: number, draggedToTaskListId: number): any;
    removeTask(taskListId: number): any;
    removeTaskCard(taskListId: number, cardId: number): any;
}

const TaskList: React.FC<ITaskListProps> = (props) => {

    const [enableCardAdd, setEnableCardAdd] = useState(false);
    const [showConfirmation, setShowConfirmation] = useState(false);
    const [taskListIdToBeRemoved, setTaskListIdToBeRemoved] = useState(0);
    const [cardIdToBeRemoved, setCardIdToBeRemoved] = useState(0);

    const addTaskCard = (e: any) => {
        setEnableCardAdd(true)
    }

    const taskCardCreated = (cardName: string, taskListId: number) => {
        props.taskCardCreated && props.taskCardCreated(taskListId, cardName);
        setEnableCardAdd(false);
    }

    const taskCardDragStart = (e: any, taskListId: number, cardId: number) => {
        e.dataTransfer.setData('taskListId', taskListId);
        e.dataTransfer.setData('cardId', cardId);
    }

    const onTaskDrop = (e: any, draggedTotaskListId: number) => {
        e.preventDefault();
        const dataTransferObj = e.dataTransfer;
        props.onTaskCardDropped && props.onTaskCardDropped(
            Number(dataTransferObj.getData('taskListId')),
            Number(dataTransferObj.getData('cardId')),
            draggedTotaskListId);
    }

    const taskCardDragOver = (e: any) => {
        e.preventDefault();
    }

    const appModalClose = () => {
        setEnableCardAdd(false);
        setShowConfirmation(false);
    }

    const removeTask = (taskListId: number) => {
        setShowConfirmation(true);
        setTaskListIdToBeRemoved(taskListId);
        setCardIdToBeRemoved(0);
    }

    const removeTaskCard = (taskListId: number, cardId: number) => {
        setShowConfirmation(true);
        setTaskListIdToBeRemoved(taskListId);
        setCardIdToBeRemoved(cardId);
    }

    const onConfirmatonOk = () => {
        setShowConfirmation(false);
        if(taskListIdToBeRemoved && cardIdToBeRemoved) {
            props.removeTaskCard && props.removeTaskCard(taskListIdToBeRemoved, cardIdToBeRemoved);
        } else if(taskListIdToBeRemoved && !cardIdToBeRemoved) {
            props.removeTask && props.removeTask(taskListIdToBeRemoved);
        }
    }

    const onConfirmatonNo = () => {
        setShowConfirmation(false);
    }

    return (
        <React.Fragment>
            <div className="task-list" onDrop={(event) => onTaskDrop(event, props.taskListId)} onDragOver={(event) => taskCardDragOver(event)}>
                <h5>{props.name}</h5>
                <RemoveIcon remove={() => removeTask(props.taskListId)} title="Remove this task"/>
                <hr />
                {
                    props.taskCard && props.taskCard.length > 0 && props.taskCard.map((card: any) => (
                        <div className="row margin-top-bottom-10px"
                            key={card.id}
                            draggable="true" onDragStart={(event) => taskCardDragStart(event, props.taskListId, card.id)}
                        >
                            <div className="col-12">
                                <div className="task-card">
                                    <div>
                                        <RemoveIcon remove={() => removeTaskCard(props.taskListId, card.id)} title="Remove this task card"/>
                                    </div>
                                    <p>{card.name}</p>
                                </div>
                            </div>
                        </div>
                    ))
                }
                <Button
                    clicked={(event) => addTaskCard(event)}
                    cssClassName="btn"
                    type="button"
                >
                    <span><i className="fas fa-plus"></i> Add a card</span>
                </Button>
            </div>
            {
                enableCardAdd &&
                <AppModal title="Create Card" width="40%" close={() => appModalClose()}>
                    <TaskCardForm
                        taskCardCreated={(cardName) => taskCardCreated(cardName, props.taskListId)}
                    />
                </AppModal>
            }
            {
                showConfirmation &&
                <AppModal title="Are your sure ?" width="30%" close={() => appModalClose()}>
                    <div className="row margin-top-20px">
                        <div className="col-12">
                            <h6 className="text-center">
                                <Button
                                    cssClassName="btn btn-info"
                                    type="button"
                                    clicked={() => onConfirmatonOk()}
                                >Yes</Button>
                                <Button
                                    type="button"
                                    clicked={() => onConfirmatonNo()}
                                >No</Button>
                            </h6>
                        </div>
                    </div>
                </AppModal>
            }
        </React.Fragment>
    )
}

export default TaskList;