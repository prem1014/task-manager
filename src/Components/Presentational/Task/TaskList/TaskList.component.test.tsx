import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import TaskList from './TaskList.component';

describe('TaskList component test suite', () => {
    Enzyme.configure({ adapter: new Adapter() });
    let componentWrapper: any;
    const taskListProps = {
        taskListId: 123,
        name: 'To do',
        taskCard: 1245,
        taskCardCreated: jest.fn(),
        onTaskCardDropped: jest.fn(),
        removeTask: jest.fn(),
        removeTaskCard: jest.fn()
    }

    beforeEach( () => {
        componentWrapper = shallow(<TaskList {...taskListProps}/>);
    });

    it('TaskList component snapshot test', () => {
        expect(componentWrapper).toMatchSnapshot();
    });
});