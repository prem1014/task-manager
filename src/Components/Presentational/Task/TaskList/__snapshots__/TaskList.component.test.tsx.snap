// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList component test suite TaskList component snapshot test 1`] = `ShallowWrapper {}`;
